### PR TITLE
(SERVER-2550) Add functions for validating cert chains and CRLs

### PR DESCRIPTION
**Pull request by mwaggett**
see individual commit messages for details

### CODE REVIEW

File: test/puppetlabs/ssl_utils/core_test.clj [hunk 1]
```diff
@@ -550,6 +550,22 @@
...
+            not-before (generate-not-before-date)
+            not-after (generate-not-after-date)
+            leaf-cert (sign-certificate issuer-name private-key serial not-before
+                                        not-after leaf-name leaf-public-key
+                                        (create-ca-extensions leaf-name serial leaf-public-key))]
+        (testing "when CRL was issued by cert in chain"
+          (is (true? (crl-issued-by-cert-in-chain? crl [leaf-cert cert]))))
+        (testing "when CRL was not issued by cert in chain"
+          (is (false? (crl-issued-by-cert-in-chain? crl [leaf-cert]))))))
+
```
Comment (mwaggett) [inline test/puppetlabs/ssl_utils/core_test.clj:568]: wasn't sure how useful this test is either. the setup is potentially unnecessary, but figured it made more sense to test with certs in the chain, rather than an empty chain.

Comment (Magisus) [reply test/puppetlabs/ssl_utils/core_test.clj:568]: Agreed. I think it's good to have.

**Review by Magisus — APPROVED** [review]

File: src/clojure/puppetlabs/ssl_utils/core.clj [hunk 2]
```diff
@@ -587,6 +587,16 @@
   (= (.getIssuerX500Principal crl)
      (.getSubjectX500Principal cert)))
 
+(schema/defn ^:always-validate crl-issued-by-cert-in-chain? :- schema/Bool
+  "Given a CRL and a certificate chain, determine whether the CRL was issued by
+  any certificate in the chain."
+  [crl :- X509CRL
+   cert-chain :- [X509Certificate]]
+  (->> cert-chain
+       (filter (partial crl-issued-by-cert? crl))
```
Comment (pcarlisle) [inline src/clojure/puppetlabs/ssl_utils/core.clj:596]: could use `some` here instead of needing to have `seq` and `boolean` after filter

**Review by Magisus — APPROVED** [review]

File: src/clojure/puppetlabs/ssl_utils/core.clj (hunk 2 above)
Comment (mwaggett) [reply src/clojure/puppetlabs/ssl_utils/core.clj:596]: ahh thanks, I figured there was a better way, but never went back after reworking this. will update.

Comment (mwaggett) [reply src/clojure/puppetlabs/ssl_utils/core.clj:596]: hmm looks like I still need to call `boolean` because `some` returns `nil` if there's no match. I guess I could just remove the output schema, but I'd kinda rather not.. Thoughts?

Comment (pcarlisle) [reply src/clojure/puppetlabs/ssl_utils/core.clj:596]: This is a place where clojure idiom kind of clashes with the schema, but yeah I'd go with the schema and keep the `boolean` call.

Comment (mwaggett) [reply src/clojure/puppetlabs/ssl_utils/core.clj]: okay cool, updated to use `some` with `boolean`

File: src/clojure/puppetlabs/ssl_utils/core.clj [hunk 3]
```diff
@@ -587,6 +587,28 @@
...
+  any certificate in the chain."
+  [crl :- X509CRL
+   cert-chain :- [X509Certificate]]
+  (->> cert-chain
+       (some (partial crl-issued-by-cert? crl))
+       boolean))
+
+(schema/defn ^:always-validate verify-crls-against-cert-chain
+  "Given a CRL chain and a certificate chain, ensure all CRLs were issued by a
+  certificate in the chain, otherwise throw."
```
Comment (Magisus) [inline src/clojure/puppetlabs/ssl_utils/core.clj:601]: Did we say we need to allow the case where people are submitting a CRL chain with other random stuff in it? If that's the case, maybe the logic we want is rather "For each of these certs, make sure there is a CRL in this chain."

Comment (justinstoller) [reply src/clojure/puppetlabs/ssl_utils/core.clj:601]: I think what we want to do is find the issuer chain of our cert and all their CRLs, ensuring those CRLs are valid and that none of the CAs in our chain of issuers has been revoked.

File: test/puppetlabs/ssl_utils/core_test.clj [hunk 4]
```diff
@@ -500,6 +501,60 @@
...
+       ;; currently UNDETERMINED REVOCATION STATUS
+       (is (thrown-with-msg? CertPathValidatorException
+                             #"Could not determine revocation status"
+                             (validate-cert-chain cert-chain-with-crl-bad-sig crl-chain-bad-signature))))
+     (testing "rejected when cert in chain has been revoked"
+       ;; REVOKED
+       (is (thrown-with-msg? CertPathValidatorException
+                             #"Certificate has been revoked*"
+                             (validate-cert-chain cert-chain-with-revoked-cert crl-chain-cert-revoked))))
+     (testing "delta CRLs are rejected"))))
```
Comment (mwaggett) [inline test/puppetlabs/ssl_utils/core_test.clj:557]: I think I'm gonna remove this - rejecting delta CRLs seems better suited for the CA service rather than having it baked into cert chain validation.

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java [hunk 5]
```diff
@@ -438,6 +445,60 @@ public static X509CRL revokeMultiple(X509CRL crl,
...
+               InvalidAlgorithmParameterException, KeyStoreException,
+               NoSuchAlgorithmException, NoSuchProviderException {
+        // there are bouncycastle versions of CertificateFactory and CertPath
+        // that I haven't looked into - not sure when using bouncycastle vs. java
+        // makes sense
+        final CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+        final CertPath certPath = certFactory.generateCertPath(certs);
+        final CertPathValidator validator;
+        if (isFIPS()) {
+            validator = CertPathValidator.getInstance(PKIX_KEYMANAGER_ALGO, BOUNCYCASTLE_JSSE_PROVIDER);
```
Comment (mwaggett) [inline src/java/com/puppetlabs/ssl_utils/SSLUtils.java:473]: this is wrong, need to figure out the appropriate algorithm/provider combo for fips.
```
java.security.NoSuchAlgorithmException: no such algorithm: PKIX for provider BCJSSE
```

File: test/puppetlabs/ssl_utils/core_test.clj [hunk 6]
```diff
@@ -500,6 +501,60 @@
...
+                             #"Could not determine revocation status"
+                             (validate-cert-chain cert-with-not-valid-crl not-yet-valid-crl))))
+     (testing "CRLs with bad signatures (aka not signed by their issuer) are rejected"
+       ;; I might want to call this something other than "bad signature",
+       ;; not sure if that's descriptive
+       ;; currently UNDETERMINED REVOCATION STATUS
+       (is (thrown-with-msg? CertPathValidatorException
+                             #"Could not determine revocation status"
+                             (validate-cert-chain cert-chain-with-crl-bad-sig crl-chain-bad-signature))))
+     (testing "rejected when cert in chain has been revoked"
```
Comment (mwaggett) [inline test/puppetlabs/ssl_utils/core_test.clj:552]: We should be able to catch this scenario when we call it in the CA service, since it has a different Reason than the other failure cases, but I think it's still appropriate to throw the error in this function.

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java [hunk 7]
```diff
@@ -438,6 +445,60 @@ public static X509CRL revokeMultiple(X509CRL crl,
...
+        } else {
+            validator = CertPathValidator.getInstance(CertPathValidator.getDefaultType());
+        }
+        final KeyStore truststore = certsToTrustStore(certs);
+        PKIXRevocationChecker revocationChecker = (PKIXRevocationChecker)validator.getRevocationChecker();
+        // we might want additional options, like SOFT_FAIL, but not sure
+        // see https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/cert/PKIXRevocationChecker.Option.html
```
Comment (Magisus) [inline src/java/com/puppetlabs/ssl_utils/SSLUtils.java:480]: I _think_ we won't need that, because we're not trying to fetch the CRL over the network (we don't use CRL distribution points).

Comment (Magisus) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java:480]: The other two options you're using here look right.

File: test/puppetlabs/ssl_utils/core_test.clj [hunk 8]
```diff
@@ -500,6 +501,60 @@
...
+     (testing "valid certs and CRL chain"
+       (is (nil? (validate-cert-chain cert-chain crl-chain))))
+     (testing "extra CRLs are okay"
+       (is (nil? (validate-cert-chain cert-chain crl-chain-extra-crls))))
+     (testing "missing CRLs are rejected"
+       ;; currently UNDETERMINED REVOCATION STATUS
+       (is (thrown-with-msg? CertPathValidatorException
+                             #"Could not determine revocation status"
+                             (validate-cert-chain cert-chain crl-chain-missing-crls))))
+     (testing "totally unrelated CRLs are rejected"
```
Comment (Magisus) [inline test/puppetlabs/ssl_utils/core_test.clj:530]: This is testing when none of the CRLs match?

Comment (mwaggett) [reply test/puppetlabs/ssl_utils/core_test.clj:530]: yeah, it's just 3 CRLs from a totally different cert chain, with different keys and everything.

Comment (mwaggett) [reply test/puppetlabs/ssl_utils/core_test.clj:530]: which sort of falls under the category of 'missing CRLs' I guess.. but it seemed conceptually different.

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 7 above)
Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java:480]: sweet

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java [hunk 9]
```diff
@@ -438,6 +445,60 @@ public static X509CRL revokeMultiple(X509CRL crl,
...
+     * @throws NoSuchProviderException
+     */
+    public static void validateCertChain(List<X509Certificate> certs,
+                                       List<X509CRL> crls)
+        throws CertificateException, CertPathValidatorException, IOException,
+               InvalidAlgorithmParameterException, KeyStoreException,
+               NoSuchAlgorithmException, NoSuchProviderException {
+        // there are bouncycastle versions of CertificateFactory and CertPath
+        // that I haven't looked into - not sure when using bouncycastle vs. java
+        // makes sense
```
Comment (justinstoller) [inline src/java/com/puppetlabs/ssl_utils/SSLUtils.java:468]: I think typically using the Java apis were available are more robust to changing BC versions or modes like FIPS vs non-FIPs. Though that idea kinda flies in the face of the NoSuchAlgorithmException you're getting below.

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 5 above)
Comment (justinstoller) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java:473]: This is weird, still looking into this, figured I should actually mention that I was doing that though ( ;

Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java:473]: okay, thanks! I was going to try using "BCFIPS" as the provider, since that was the only thing I found in my googling that seemed like it could maybe work haha. but I need Travis to test that for me, so I'll push something soon.

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 9 above)
Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java:468]: I was wondering about that, but that error is getting thrown by the `CertPathValidator` class, which doesn't seem to have a corresponding BC class, afaict.

Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java:468]: so I guess that's why

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java [hunk 10]
```diff
@@ -438,6 +444,54 @@ public static X509CRL revokeMultiple(X509CRL crl,
...
+        final KeyStore truststore = certsToTrustStore(certs);
+        PKIXRevocationChecker revocationChecker = (PKIXRevocationChecker)validator.getRevocationChecker();
+        revocationChecker.setOptions(EnumSet.of(
+                PKIXRevocationChecker.Option.PREFER_CRLS,
+                PKIXRevocationChecker.Option.NO_FALLBACK));
+        PKIXBuilderParameters params =
+                new PKIXBuilderParameters(truststore, new X509CertSelector());
+        params.addCertPathChecker(revocationChecker);
+        // tests pass when this is true, false, or not included at all..
+        //params.setRevocationEnabled(true);
```
Comment (justinstoller) [inline src/java/com/puppetlabs/ssl_utils/SSLUtils.java:488]: It looks like it enabled a value that is overridden by the `addCertPathChecker` above: https://docs.oracle.com/javase/7/docs/api/java/security/cert/PKIXParameters.html#setRevocationEnabled(boolean)

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 5 above)
Comment (justinstoller) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: Yeah, I see the BCFIPS being used in some examples. This just makes me confused because we should be configuring the BCJSSE to work in "fips" mode, ie, actually be backed by the BCFIPS provider. I don't know why BCJSSE wouldn't eventually defer to the BCFIPS provider here, or why if it doesn't do it here it would elsewhere... I have a feeling that it's an arcane java configuration issue or BC implementation quirk that would take forever to confirm.

File: test/puppetlabs/ssl_utils/core_test.clj (hunk 6 above)
Comment (justinstoller) [reply test/puppetlabs/ssl_utils/core_test.clj:546]: Seems like there was a little bit of disagreement in the ticket between stakeholders as to whether we should honor this case, too.

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 5 above)
Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: I don't really know anything about BCJSSE or providers in general - are you saying something is off somewhere that we need to fix?

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 10 above)
Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java:488]: ahhh, I swear I read that, but I must've decided I got the gist before making it all the way to the end haha

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 5 above)
Comment (justinstoller) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: My understanding is that you should be able to say that you want the BCJSSE provider and then the java.security file we use should have it configured to actually be BCFIPS provider (or regular BC provider, depending on the profile you run the tests with).

I don't know why that doesn't work, it's probably just my understanding of it that's broken and it's probably a long road to fix that...

Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: ah, interesting. so would that behavior, in theory, come from the BCJSSE provider? like, it includes magic to detect the right underlying provider?

I'm noticing that the existing uses of the BCJSSE provider are only in FIPS mode. If I'm understanding what you're saying correctly, I would expect us to use that everywhere, rather than just for FIPS? I sorta just assumed that BCJSSE was a separate, unrelated provider. BCJSSE looks sorta specific to SSL stuff? https://www.bouncycastle.org/docs/tlsdocs1.5on/org/bouncycastle/jsse/package-summary.html

Comment (justinstoller) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: I think section 2.1-3.1.2 of this explains it better than I can: https://downloads.bouncycastle.org/fips-java/BC-FJA-(D)TLSUserGuide-1.0.9.pdf

Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: I'll take a look on Monday 😜

File: test/puppetlabs/ssl_utils/core_test.clj [hunk 11]
```diff
@@ -500,6 +501,52 @@
...
+         cert-with-expired-crl (-> "crl_validation/cert-with-expired-crl.pem" open-ssl-file pem->certs)
+         expired-crl (-> "crl_validation/expired-crl.pem" open-ssl-file pem->crls)
+         cert-with-not-valid-crl (-> "crl_validation/cert-with-not-valid-crl.pem" open-ssl-file pem->certs)
+         not-yet-valid-crl (-> "crl_validation/not-yet-valid-crl.pem" open-ssl-file pem->crls)
+         cert-chain-with-crl-bad-sig (-> "crl_validation/cert-with-crl-bad-sig.pem" open-ssl-file pem->certs)
+         crl-chain-bad-signature (-> "crl_validation/crl-with-bad-signature.pem" open-ssl-file pem->crls)
+         cert-chain-with-revoked-cert (-> "crl_validation/cert-chain-with-revoked-cert.pem" open-ssl-file pem->certs)
+         crl-chain-cert-revoked (-> "crl_validation/crl-chain-with-cert-revoked.pem" open-ssl-file pem->crls)]
+     (testing "successful with single-cert chain and CRL"
+       (let [[cert crl] (generate-cert-chain-with-crls 1)]
```
Comment (mwaggett) [inline test/puppetlabs/ssl_utils/core_test.clj:520]: gonna add actual fixtures for this, just wanted to test real quick

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java [hunk 12]
```diff
@@ -465,30 +464,24 @@ public static void validateCertChain(List<X509Certificate> certs,
...
         final KeyStore truststore = certsToTrustStore(certs);
-        PKIXRevocationChecker revocationChecker = (PKIXRevocationChecker)validator.getRevocationChecker();
-        revocationChecker.setOptions(EnumSet.of(
-                PKIXRevocationChecker.Option.PREFER_CRLS,
-                PKIXRevocationChecker.Option.NO_FALLBACK));
         PKIXBuilderParameters params =
                 new PKIXBuilderParameters(truststore, new X509CertSelector());
-        params.addCertPathChecker(revocationChecker);
-        // tests pass when this is true, false, or not included at all..
-        //params.setRevocationEnabled(true);
```
Comment (justinstoller) [inline src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: when you set this to `false` does the validation succeed in the tests (and the tests then fail)?

Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: oh, I should try this again now that I got rid of the revocation checker.

Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: okay yes, now that I've removed the revocation checker, setting this to `false` makes all my test cases valid, so the tests fail.

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 5 above)
Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: > The BC JSSE provider is built on top of the TLS API and provides functionality to support the use of the JSSE API defined in javax.net.ssl

Since the things I'm adding aren't in ^that java class, I think it makes sense that the BCJSSE provider doesn't work for this.

File: src/java/com/puppetlabs/ssl_utils/SSLUtils.java (hunk 12 above)
Comment (mwaggett) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: Were you thinking I should explicitly set this to `true` now or do you think it's fine to let the default handle it?

Comment (justinstoller) [reply src/java/com/puppetlabs/ssl_utils/SSLUtils.java]: I think it's fine to leave it using the default handler, I just wanted to know that the default behavior was the same between BCFIPS' and Java's built in verifiers.

File: test/puppetlabs/ssl_utils/core_test.clj [hunk 13]
```diff
@@ -500,6 +502,59 @@
             (is (revoked? updated-crl cert))
             (is (revoked? updated-crl cert2))))))))
 
+(deftest validate-cert-chain-and-crls-test
+ (testing "validate cert chain against CRLs"
```
Comment (mwaggett) [inline test/puppetlabs/ssl_utils/core_test.clj:506]: I realize I didn't add test cases around invalid certs (just invalid CRLs). While that was originally motivated by my needs for this ticket (since the incoming CRLs are the only piece we won't have control over), I'm thinking it might be okay, since we'd kinda just be testing the Java libs we're using? Thoughts?

File: test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem [hunk 14]
```diff
@@ -0,0 +1,57 @@
+-----BEGIN CERTIFICATE-----
```
Comment (Magisus) [inline test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem:1]: Was this needed to generate the `other-crl-chain`? Might be worth not keeping these around as a file, since they seem to be unused?

Comment (Magisus) [reply test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem:1]: Seeing this gave me the idea to double check that the `other-crl-chain` is actually valid when used with its own certs, and it is. That being said, thinking about it more, I do still feel like the old error ("No CRLs for issuer") for that one test made more sense... You never figured out what's going on there?

**Review by Magisus — COMMENTED** [review]
I think this generally looks good to me, but I obviously haven't dived very deep into it.

File: test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem (hunk 14 above)
Comment (mwaggett) [reply test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem:1]: yeah, this was generated along with the `other-crl-chain`. I figured it might be useful to have the certs that generated those CRLs, but I'm happy to remove this if we don't think we need it.

But yeah, I still don't really know what's going on with the changed error message.. I did figure out that it's actually the `3-cert/crl-chain` that changed somewhere though, not the `other-*` ones. I could keep going back farther to try and figure out the state I was in when those were originally generated, but I'm not even positive that was a state that ultimately got committed. I tried to regenerate whenever I made updates, but since I was originally developing on non-FIPS and this message change only shows itself in FIPS mode, I could've easily squashed some state that I thought didn't change the behavior.

File: test/puppetlabs/ssl_utils/core_test.clj [hunk 15]
```diff
@@ -500,6 +502,59 @@
...
+                                                          #"Could not determine revocation status")
+         revoked-cert-error-message (if (SSLUtils/isFIPS) #"Certificate revocation after*"
+                                                          #"Certificate has been revoked*")]
+     (testing "successful with single-cert chain and CRL"
+       (is (nil? (validate-cert-chain single-cert-chain single-crl-chain))))
+     (testing "successful when all certs have an associated CRL"
+       (is (nil? (validate-cert-chain cert-chain crl-chain))))
+     (testing "successful with extra CRLs"
+       (is (nil? (validate-cert-chain cert-chain crl-chain-extra-crls))))
+     (testing "fails with missing CRLs"
```
Comment (mwaggett) [inline test/puppetlabs/ssl_utils/core_test.clj:534]: mmmm okay so this doesn't actually fail if just the leaf CRL is missing.. trying to decide if I think that behavior belongs in this function or not, since, as-is, it could be used to verify any cert chain, as opposed to a CA chain specifically. kinda leaning towards not.. Thoughts?

Comment (Magisus) [reply test/puppetlabs/ssl_utils/core_test.clj:534]: I think that case makes sense, since we don't actually have to check that CRL for anything. If your chain had a non-CA cert as the leaf, and the leaf CRL was missing, I assume it _would_ break. I'm indifferent to whether we test that here.

Comment (mwaggett) [reply test/puppetlabs/ssl_utils/core_test.clj:534]: Since the leaf CRL comes from _our_ intermediate CA, as opposed to coming from users, I'm not so worried about it being missing.

File: test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem (hunk 14 above)
Comment (Magisus) [reply test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem:1]: Okay.

File: test/puppetlabs/ssl_utils/core_test.clj (hunk 15 above)
Comment (Magisus) [reply test/puppetlabs/ssl_utils/core_test.clj:534]: This is a general-purpose function, but all of the testing here is around the cases where every element of the chain is a CA cert. In that context, it makes sense to not test the leaf CRL, since it will never be relevant to validating chains like that.

Comment (mwaggett) [reply test/puppetlabs/ssl_utils/core_test.clj:534]: right. I think I've had iterations of the `missing-crls` file where the root CRL was missing and where an intermediate CRL was missing and they both fail as we expect. It's just the leaf CRL that it doesn't seem to care about, which makes sense for the reasons you said.

File: test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem (hunk 14 above)
Comment (mwaggett) [reply test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem:1]: @Magisus did you want me to remove this file?

Comment (Magisus) [reply test-resources/puppetlabs/ssl_utils/examples/ssl/crl_validation/other-3-cert-chain.pem:1]: Mmm... no, thinking about it more, I guess it's nice to have the certs associated with those CRLs, like you were saying. Disregard that comment 🙂

**Review by Magisus — APPROVED** [review]
